pkg/backend/httpstate/cloud_registry.go

```diff
@@ -57,3 +59,36 @@ func (r *cloudRegistry) GetPackage(
 	}
 	return meta, err
 }
+
+func (r *cloudRegistry) PublishTemplate(ctx ctx.Context, op backend.TemplatePublishOp) error {
+	startResp, err := r.cl.StartTemplatePublish(ctx, op.Source, op.Publisher, op.Name, op.Version)
+	if err != nil {
+		return fmt.Errorf("failed to initialize template publish: %w", err)
+	}
+
+	uploadURL := startResp.UploadURLs.Archive
+	req, err := http.NewRequestWithContext(ctx, "PUT", uploadURL, op.Archive)
+	if err != nil {
+		return fmt.Errorf("failed to create upload request: %w", err)
+	}
+
+	req.Header.Set("Content-Type", "application/gzip")
+
+	httpClient := &http.Client{}
+	resp, err := httpClient.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed to upload archive: %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusNoContent {
+		return fmt.Errorf("upload failed with status %d", resp.StatusCode)
+	}
+
+	err = r.cl.CompleteTemplatePublish(ctx, op.Source, op.Publisher, op.Name, op.Version.String(), startResp.OperationID)
+	if err != nil {
+		return fmt.Errorf("failed to complete template publish: %w", err)
+	}
+
+	return nil
+}
```

Comment:
The logic here should be in `pkg/backend/httpstate/client/`. See `Client.PublishPackage` for an example to mirror:

https://github.com/pulumi/pulumi/blob/f6f2e559d82cd7d88b519d64b65f4281d78d0dd4/pkg/backend/httpstate/client/client.go#L1493